packages/relay-runtime/store/observeFragmentExperimental.js

```diff
@@ -173,6 +170,67 @@ function observeSelector<TFragmentType: FragmentType, TData>(
   });
 }
 
+function observePluralSelector<
+  TFragmentType: FragmentType,
+  TData: Array<mixed>,
+>(
+  environment: IEnvironment,
+  fragmentNode: Fragment<TFragmentType, TData>,
+  fragmentSelector: PluralReaderSelector,
+): Observable<FragmentState<TData>> {
+  const snapshots = fragmentSelector.selectors.map(selector =>
+    environment.lookup(selector),
+  );
+
+  return Observable.create(sink => {
+    const states = snapshots.map((snapshot, index) =>
+      snapshotToFragmentState(
+        environment,
+        fragmentNode,
+        fragmentSelector.selectors[0].owner,
+        snapshot,
+      ),
+    );
+
+    const reduceState = () =>
+      states.reduce<FragmentState<TData>>(
+        (acc, state) => {
+          if (acc.state !== 'ok') {
+            return acc;
```

Comment:
I wonder if this would be easier to read as a for loop with a `break`? It's a little hard to follow right now, and I think right now we still have to continue iterating the states even after we hit one that's not "ok". Would be nicer if we could just bail and return the loading or error state right away.

Maybe we could define a generic function (outside of this closure) which does this transformation:

```ts
function namingIsHard<T>(states: Array<FragmentState<T>>): FragmentState<Array<T>> {
  const values: Array<T> = [];
  for(let i = 0; i < states.length; i++) {
    const state = states[i];
    if(state.state === 'ok') {
      values.push(state.value);
    } else {
      return state;
    }
  }
  return { state: "ok", value: values };
}
```

Reply:
https://github.com/facebook/relay/pull/4862/commits/4a5629881cafb7d8b53fc9d18314b9afe875f749